### PR TITLE
dev-cmd/generate-cask-api: start generating cask internal JSON v3

### PR DESCRIPTION
**Pull request by apainintheneck**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Part of https://github.com/Homebrew/brew/issues/16410

This adds a new file to the output of `brew generate-cask-api` which represents the new internal JSON v3 file. It involves removing a bunch of unneeded hash keys while removing blank ones as well.

I've made some slight changes to the cask loader as well but more might be necessary before this starts loading things correctly. The full loader code will be added in a separate PR.

### Skippable hash fields
- appcast
  - This is disabled right now and no longer used in the core cask tap.
- deprecated
  - Not used
- disabled
  - Not used
- full_token
  - Not used
- tap
  - Will be included in tap api hash
- tap_git_head
  - Will be included in tap api hash
- ruby_source_checksum
  - Replaced by ruby_source_sha256 to match formula API hash

### Nilable hash fields
- languages
- container
- deprecation_date
- deprecation_reason
- disable_date
- disable_reason
- caveats
- depends_on
- conflicts_with
- url_specs

### New hash fields
- caskfile_only
  - allows us to reduce the size of the artifacts array
- ruby_source_sha256
  - to match formula API hash

### Updated cask loader
- to handle new ruby_source_sha256 field

## Post Merge
- I'll open a PR to start signing this new file in formulae.brew.sh similar to this other PR for formulae: https://github.com/Homebrew/formulae.brew.sh/pull/1069

### CODE REVIEW

**Comment by apainintheneck** [comment]
Before:

```
-rw-r--r--  1 kevinrobell  staff   6.5M Mar  2 15:29 cask.jws.json
```

After:

```
-rw-r--r--  1 kevinrobell  wheel   4.2M Mar  2 17:38 homebrew-cask.json
```

File: Library/Homebrew/cask/cask.rb [hunk 1]
```diff
@@ -283,7 +283,10 @@ def populate_from_api!(json_cask)
       @tap_git_head = json_cask.fetch(:tap_git_head, "HEAD")
 
       @ruby_source_path = json_cask[:ruby_source_path]
-      @ruby_source_checksum = json_cask[:ruby_source_checksum].freeze
+
+      ruby_source_sha256 = json_cask.dig(:ruby_source_checksum, :sha256)
+      ruby_source_sha256 ||= json_cask[:ruby_source_sha256]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:288]: What do you think about either `# TODO`ing or splitting into separate functions these paths so they don't outlast the move to non-internal JSON?

File: Library/Homebrew/cask/cask.rb [hunk 2]
```diff
@@ -347,15 +342,64 @@ def to_h
       }
     end
 
-    def to_hash_with_variations
-      if loaded_from_api? && !Homebrew::EnvConfig.no_install_from_api?
-        return api_to_local_hash(Homebrew::API::Cask.all_casks[token].dup)
+    # @private
+    def to_api_hash
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/cask.rb:346]:
```suggestion
    def to_internal_api_hash
```

assuming this is actually correct

File: Library/Homebrew/tap.rb [hunk 3]
```diff
@@ -1309,6 +1309,22 @@ def tap_migrations
       migrations
     end
   end
+
+  sig { returns(T::Hash[String, T.untyped]) }
+  def to_api_hash
```
Comment (MikeMcQuaid) [inline Library/Homebrew/tap.rb:1314]:
```suggestion
  def to_internal_api_hash
```

again, if correct.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good! A few non-blocking thoughts/comments.

File: Library/Homebrew/cask/cask.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:346]: I'll just rename the formula ones too. I have no preference personally but I guess it indicates intent a bit better than before.

File: Library/Homebrew/cask/cask.rb (hunk 1 above)
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb:288]: Good idea. It is quite ugly right now.

File: Library/Homebrew/cask/cask.rb (hunk 2 above)
Comment (apainintheneck) [reply Library/Homebrew/cask/cask.rb]: Addressed in https://github.com/Homebrew/brew/pull/16798/commits/b5c9651b4624a5349202c3346c5983cd7e27ac5e

**Comment by apainintheneck** [comment]
I'll merge this in tomorrow after I finish testing some things to make doubly sure that I'm not breaking current API JSON generation.

**Review by MikeMcQuaid — APPROVED** [review]
Looking good, thanks for making those changes ❤️

**Comment by apainintheneck** [comment]
Waiting on discussion in https://github.com/Homebrew/brew/pull/16826#issuecomment-1985027029

**Pull request merged**